snare/tests/test_get_body.py

```diff
@@ -0,0 +1,83 @@
+import unittest
+import aiohttp
+import yarl
+import sys
+import os
+import shutil
+import asyncio
+from snare.cloner import Cloner
+from snare.utils.asyncmock import AsyncMock
+from snare.utils.page_path_generator import generate_unique_path
+
+
+class TestGetBody(unittest.TestCase):
+    def setUp(self):
+        self.main_page_path = generate_unique_path()
+        os.makedirs(self.main_page_path)
+        self.root = 'http://example.com'
+        self.level = 0
+        self.max_depth = sys.maxsize
+        self.loop = asyncio.new_event_loop()
+        self.css_validate = 'false'
+        self.handler = Cloner(self.root, self.max_depth, self.css_validate)
+        self.target_path = '/opt/snare/pages/{}'.format(yarl.URL(self.root).host)
+        self.return_content = None
+        self.expected_content = None
+        self.filename = None
+        self.hashname = None
+        self.url = None
+        self.content = b'''<html><body><a href="http://example.com/test"></a></body></html>'''
+
+        self.session = aiohttp.ClientSession
+        self.session.get = AsyncMock(
+            return_value=aiohttp.ClientResponse(
+                url=yarl.URL("http://www.example.com"), method="GET", writer=None, continue100=1,
+                timer=None, request_info=None, traces=None, loop=self.loop,
+                session=None
+            )
+        )
+
+    def test_get_body(self):
+        aiohttp.ClientResponse._headers = {'Content-Type': 'text/html'}
+        aiohttp.ClientResponse.read = AsyncMock(return_value=self.content)
+        self.filename, self.hashname = self.handler._make_filename(yarl.URL(self.root))
+
+        async def test():
+            await self.handler.new_urls.put((yarl.URL(self.root), 0))
+            await self.handler.get_body(self.session)
+
+        self.loop.run_until_complete(test())
+        with open(os.path.join(self.target_path, self.hashname)) as f:
+            self.return_content = f.read()
+
+        self.assertEqual(self.return_content, '<html><body><a href="/test"></a></body></html>')
+
+    def test_get_body_css_validate(self):
+        aiohttp.ClientResponse._headers = {'Content-Type': 'text/css'}
+
+        self.css_validate = 'true'
+        self.handler = Cloner(self.root, self.max_depth, self.css_validate)
+        self.content = b'''.banner { background: url("/example.png") }'''
```

Comment:
Check for other css entities like [this](https://github.com/mushorg/snare/blob/master/snare/cloner.py#L154) and some others.

Reply:
Please check the global variable: `self.visited_urls`, `self.meta` , `self.new_urls`. I won't comment for each test case, just check them where it is necessary.